tslearn/shapelets.py

```diff
@@ -530,16 +568,72 @@ def locate(self, X):
                [0]])
         """
         check_is_fitted(self, '_X_fit_dims')
-        X = check_array(X, allow_nd=True)
-        X = to_time_series_dataset(X)
-        X = check_dims(X, X_fit_dims=self._X_fit_dims)
+        X = check_array(X, allow_nd=True, force_all_finite=False)
+        X = self._preprocess_series(X)
+        X = check_dims(X, X_fit_dims=self._X_fit_dims,
+                       check_n_features_only=True)
+        self._check_series_length(X)
+
         n_ts, sz, d = X.shape
         locations = self.locator_model_.predict(
             [X[:, :, di].reshape((n_ts, sz, 1)) for di in range(self.d_)],
             batch_size=self.batch_size, verbose=self.verbose
         )
         return locations.astype(numpy.int)
 
+    def _check_series_length(self, X):
+        """Ensures that time series in X matches the following requirements:
+        
+        - their length is greater than the size of the longest shapelet
+        - (at predict time) their length is lower than the maximum allowed 
+        length, as set by self.max_size
+        """
+        sizes = numpy.array([ts_size(Xi) for Xi in X])
+
+        if self.n_shapelets_per_size is not None:
+            max_sz_shapelets = max(self.n_shapelets_per_size.keys())
+            min_sz_X = sizes.min()
+            if max_sz_shapelets > min_sz_X:
+                raise ValueError("Sizes in X do not match maximum "
+                                 "shapelet size: there is at least one "
+                                 "series in X that is shorter than one of the "
+                                 "shapelets. Shortest time series is of "
+                                 "length {} and longest shapelet is of length "
+                                 "{}".format(min_sz_X, max_sz_shapelets))
+
+        if hasattr(self, 'model_') or self.max_size is not None:
+            # Model is already fitted
+            max_sz_X = sizes.max()
+
+            if hasattr(self, 'model_'):
+                max_size = self._X_fit_dims[1]
+            else:
+                max_size = self.max_size
+            if max_size < max_sz_X:
+                raise ValueError("Sizes in X do not match maximum allowed "
+                                 "size as set by max_size. "
+                                 "Longest time series is of "
+                                 "length {} and max_size is "
+                                 "{}".format(max_sz_X, self.max_size))
+
+    def _preprocess_series(self, X):
+        if self.scale:
+            X = TimeSeriesScalerMinMax().fit_transform(X)
+        else:
+            X = to_time_series_dataset(X)
+        if self.max_size is not None and self.max_size != X.shape[1]:
```

Comment:
Shouldn't this be `<=` instead of `!=` ?

Reply:
Well:

* if `self.max_size == X.shape[1]` then we have no problem
* if `X.shape[1] > self.max_size` then we raise an error

so the only real reshaping is done when `self.max_size < X.shape[1]`, which seems ok, no?

Reply:
Ok yes I looked over that if on the next line :)